src/Symfony/Component/Serializer/Normalizer/ArrayDenormalizer.php

```diff
@@ -72,8 +82,15 @@ public function supportsDenormalization($data, string $type, string $format = nu
             throw new BadMethodCallException(sprintf('The nested denormalizer needs to be set to allow "%s()" to be used.', __METHOD__));
         }
 
-        return str_ends_with($type, '[]')
-            && $this->denormalizer->supportsDenormalization($data, substr($type, 0, -2), $format, $context);
+        if (!str_ends_with($type, '[]')) {
+            return false;
+        }
+
+        if ($this->checkFirstElement) {
+            $data = 0 === \count($data) ? null : reset($data);
```

Comment:
`$data = $data ? reset($data) : null;`

but are we sure this is always an array?

Reply:
Was thinking we could assume an array was given in case the `$type` variable ends with `[]`. But it might be better to be preventive about the `$data` value.

In the issue some ideas came across that the `supportsDenormalization(...)` should not check the value of the `$data` parameter. instead the `denormalize(...)` should perform the check and fail in case it's not met with the requirements. 

So if we want to comply with that idea we can't do `return false;` incase `$data` is not an array. So I was thinking  too replace the `$data` parameter with NULL when the parameter doesn't contain an array (just like when the array is empty). And let the `denormalize(...)` fail of not having an array as `$data`.

Will this be an acceptable solution?

Reply:
An `is_array` check would be in place I guess, to cover the (invalid) cases where not an array is passed in. Otherwise a TypeError will be thrown on `count`.

Reply:
In case `$data` is not an array, I replace it with an array so the behind it will act as if the array is empty. Don't think this is the solution because @derrabus mentioned another issue with passing NULL to denormalizers. But will solve this conversation and let's continue on the other.